kimu/core/intersection_tool_line_circle.py

```diff
@@ -0,0 +1,272 @@
+            - 2.0 * (line_coords[0]) ** 2.0 * centroid[1] * line_coords[3]
+            + 2.0 * centroid[1] * line_coords[1] * line_coords[2] * line_coords[0]
+            + (line_coords[2]) ** 2.0 * (centroid[1]) ** 2.0
+            - (line_coords[2]) ** 2.0 * r ** 2.0
+            - 2.0 * line_coords[0] * line_coords[2] * (centroid[1]) ** 2.0
+            + 2.0 * line_coords[0] * line_coords[2] * r ** 2.0
+            + (line_coords[0]) ** 2.0 * (centroid[1]) ** 2.0
+            - (line_coords[0]) ** 2.0 * r ** 2.0
+        )
+        # print(f"Value of c is {c}")
+
+        # Check that the selected line feature and indirectly
+        # defined circle intersect.
+        # To do: Complains that there is no intersection point
+        # of given radius if radius is small (e.g. 10m)?
+        # The limit value seems to change according to coordinates
+        # of centroid point..
+        sqrt_in = b ** 2.0 - 4.0 * a * c
+        print(f"Value of sqrt_in is {sqrt_in}")
+        if sqrt_in < 0.0 or a == 0.0:
+            LOGGER.warning(
+                tr("There is no intersection point(s)!"),
+                extra={"details": ""},
+            )
+            return
+
+        # Computing the coordinates for intersection points
+        x_sol1 = (-b + math.sqrt(sqrt_in)) / (2.0 * a)
+
+        y_sol1 = (
+            x_sol1 * line_coords[3]
+            - line_coords[0] * line_coords[3]
+            - x_sol1 * line_coords[1]
+            + line_coords[2] * line_coords[1]
+        ) / (line_coords[2] - line_coords[0])
+
+        x_sol2 = (-b - math.sqrt(sqrt_in)) / (2.0 * a)
+
+        y_sol2 = (
+            x_sol2 * line_coords[3]
+            - line_coords[0] * line_coords[3]
+            - x_sol2 * line_coords[1]
+            + line_coords[2] * line_coords[1]
+        ) / (line_coords[2] - line_coords[0])
+
+        # Check that the result point lies in the map canvas extent
+        extent = iface.mapCanvas().extent()
+
+        if (
+            x_sol1 < extent.xMinimum()
+            or x_sol1 > extent.xMaximum()
+            or y_sol1 < extent.yMinimum()
+            or y_sol1 > extent.yMaximum()
+        ):
+            LOGGER.warning(
+                tr("Intersection point 1 lies outside of the map canvas!"),
+                extra={"details": ""},
+            )
+            return
+
+        if (
+            x_sol2 < extent.xMinimum()
+            or x_sol2 > extent.xMaximum()
+            or y_sol2 < extent.yMinimum()
+            or y_sol2 > extent.yMaximum()
+        ):
+            LOGGER.warning(
+                tr("Intersection point 2 lies outside of the map canvas!"),
+                extra={"details": ""},
+            )
+            return
+
+        # Check that the line genuinely intersects with a circle.
+        # In case the line only touches the circle, only one result
+        # layer gets generated since x_sol1 = x_sol2
+        if sqrt_in == 0.0:
```

Comment:
This is quite a long method, could use refactoring. E.g. maybe move the contents of the final if-else blocks to their own methods, separate the a-b-c steps above to their own method, ...

Reply:
I see.. I wouldn't separate the computation of parameter values a, b and c since they are all related to the same solution formula, but I could separate all of them into one separate method. I also could make the contents of the last if-else structure into two separate methods and just call them if needed (depending on whether the line genuinely intersects with the circle or not)